best-first-cv/script.whizzml

```diff
@@ -0,0 +1,561 @@
+                          "regularization"
+                          "seed"])
+(define EVALUATION_OPTIONS ["sample_rate"
+                            "input_fields"
+                            "out_of_bag"
+                            "range"
+                            "replacement"
+                            "ordering"
+                            "seed"
+                            "missing_strategy"
+                            "combiner"])
+;; cross-validation
+;;
+;; creates k-fold cross-validation for a dataset
+;; Inputs:
+;;   dataset-id: (string) Dataset ID
+;;   k-folds: (integer) Number of folds
+;;   model-options: (map) Options to use in model/ensemble
+;;   evaluation-options: (map) Options to use in evaluation creation
+;;   delete-resources: (boolean) Whether to delete all intermediate resources
+;;
+;; Output: (map) Average of evaluations results
+;;
+;; Raises:
+;;  101: The dataset-id argument is not a string
+;;  102: The dataset-id is not a valid dataset ID
+;;  103: The k-folds argument is not an integer
+;;  104: The k-folds argument is not >= 2
+;;  106: The objective field ID is not in the selectable IDs list
+;;
+(define (cross-validation k-fold-datasets
+                          objective-id
+                          model-options
+                          evaluation-options
+                          delete-resources)
+  (let (dataset (fetch (k-fold-datasets 0))
+        dataset-name (dataset "name" false))
+    (check-dataset-objective-id objective-id dataset)
+    (let (objective-name (get-objective-name dataset objective-id)
+          evaluations (create-k-evaluations k-fold-datasets
+                                            objective-name
+                                            dataset-name
+                                            model-options
+                                            evaluation-options
+                                            delete-resources)
+          ;; commented out till wintermute's new version
+          ;; evaluations-average (create-and-wait-evaluation {"evaluations" evaluations})
+          average-phi (get-average-phi evaluations)
+          _ (log-info "model options: " model-options)
+          _ (log-info "average phi: " average-phi))
+      (when delete-resources
+        (map safe-delete evaluations))
+      average-phi)))
+
+
+(define (safe-create type args)
+  (try
+    (create type args)
+    (catch e
+      (create type args))))
+
+(define (get-average-phi evaluations)
+  (/ (iterate (acc 0 ev evaluations)
+       (+ acc ((fetch ev) ["result" "model" "average_phi"])))
+     (count evaluations)))
+
+;; check-resource-id
+;;
+;; Validates that the argument is a resource ID and its type. Raises an error
+;; if otherwise.
+;;
+;; Inputs:
+;;   resource-id: (string) Resource ID
+;;   type: (string) Type of resource
+;;
+;; Output: (string) Checked resource ID
+(define (check-resource-id resource-id type)
+  (when (not (string? resource-id))
```

Comment:
there is a whizzml standard procedure `resource-id?` that also checks that the string is well formed, so here you can say: `(when (not (resource-id? resource-id)`